### PR TITLE
feat(changelog)!: add postprocessors

### PR DESCRIPTION
**Pull request by PSeitz**
## Description

Allow same transformations on the final output as with the git commit preprocessor.
This will enable things like replacing usernames with github usernames #132

## How Has This Been Tested?

tested locally

## Types of Changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation (no code change)
- [ ] Refactor (refactoring production code)
- [ ] Other 

## Checklist:

- [ ] My code follows the code style of this project.
- [x] I have updated the documentation accordingly.
- [x] I have formatted the code with [rustfmt](https://github.com/rust-lang/rustfmt).
- [x] I checked the lints with [clippy](https://github.com/rust-lang/rust-clippy).
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

**Comment by orhun** [comment]
Hello, thanks for the contribution! 🐻 

Can you maybe provide an example about this feature in the documentation? It is not clear what changes are made from the code.

**Comment by orhun** [comment]
Also, isn't this a library-wise breaking change since `CommitPreprocessor` is renamed as `TextPreprocessor`?

File: git-cliff-core/src/changelog.rs [hunk 1]
```diff
@@ -142,7 +143,29 @@ impl<'a> Changelog<'a> {
 			write!(out, "{header}")?;
 		}
 		for release in &self.releases {
-			write!(out, "{}", self.template.render(release)?)?;
+			let mut rendered = self.template.render(release)?;
+			if let Some(preprocessors) =
+				self.config.changelog.postprocessors.as_ref()
+			{
+				for preprocessor in preprocessors {
+					if let Some(text) = &preprocessor.replace {
```
Comment (orhun) [inline git-cliff-core/src/changelog.rs:151]: Please add tests to cover this change.

Comment (PSeitz) [reply git-cliff-core/src/changelog.rs:151]: I added a test

**Comment by PSeitz** [comment]
> Also, isn't this a library-wise breaking change since `CommitPreprocessor` is renamed as `TextProcessor`?

Yes, I didn't know it's also used as a library.

I updated the docs with the new feature.

**Comment by PSeitz** [comment]
Fixed the commit message by squashing

**Comment by PSeitz** [comment]
The test failure is unrelated

File: website/docs/configuration.md [hunk 2]
```diff
@@ -56,6 +57,12 @@ It is useful for adding indentation to the template for readability, as shown [i
 
 Footer text that will be added to the end of the changelog.
 
+### postprocessors
+
+An array of commit preprocessors for manipulating the changelog before outputting.
```
Comment (orhun) [inline website/docs/configuration.md:62]: Shouldn't that be "An array of commit postprocessors"?

File: website/docs/configuration.md [hunk 3]
```diff
@@ -56,6 +57,12 @@ It is useful for adding indentation to the template for readability, as shown [i
...
+Can e.g. be used for replacing commit author with github user names.
```
Comment (orhun) [inline website/docs/configuration.md:63]: Nit: `github` -> `GitHub`

File: git-cliff-core/src/changelog.rs [hunk 4]
```diff
@@ -142,7 +143,29 @@ impl<'a> Changelog<'a> {
...
+						if preprocessor.pattern.is_match(&rendered) {
+							rendered = command::run(
+								command,
+								Some(rendered.to_string()),
+								vec![],
+							)?;
+						}
+					}
+				}
+			}
```
Comment (orhun) [inline git-cliff-core/src/changelog.rs:166]: I see that this is simply the same method  as `Commit::preprocess`. Can you refactor this so there isn't code duplication? One idea would be adding an helper method to somewhere, possibly in `TextProcessor` about this.

File: git-cliff-core/src/config.rs [hunk 5]
```diff
@@ -29,13 +29,15 @@ pub struct Config {
...
 	/// Changelog body, template.
-	pub body:   Option<String>,
+	pub body:           Option<String>,
 	/// Changelog footer.
-	pub footer: Option<String>,
+	pub footer:         Option<String>,
 	/// Trim the template.
-	pub trim:   Option<bool>,
+	pub trim:           Option<bool>,
+	/// Changelog postrocessors.
```
Comment (orhun) [inline git-cliff-core/src/config.rs:39]: Nit: `postrocessors` -> `postprocessors`

**Review by orhun — CHANGES_REQUESTED** [review]
Had some comments / suggestions.

**Comment by orhun** [comment]
Can you also check out #165 and pre-empt your opinion before this PR is merged?

File: git-cliff-core/src/changelog.rs (hunk 4 above)
Comment (PSeitz) [reply git-cliff-core/src/changelog.rs]: I merged them

**Review by orhun — APPROVED** [review]

**Pull request merged**